internal/jobworker/cgroup/controller.go

```diff
@@ -0,0 +1,215 @@
+}
+
+// cpuController enables and applies the "cpu.max" control.
+type cpuController struct {
+	baseController
+	cpus float32
+}
+
+func (c cpuController) apply() error {
+	const (
+		period = 100000
+	)
+	limit := c.cpus * period
+	value := fmt.Sprintf("%f %d", limit, period)
+
+	return errors.WithStack(c.baseController.apply(cpuMax, value))
+}
+
+// newMemoryController creates a memoryController instance.
+func newMemoryController(cgroup Cgroup, limit uint64) *memoryController {
+	return &memoryController{
+		baseController: baseController{name: memory, cgroup: cgroup},
+		limit:          limit,
+	}
+}
+
+// memoryController enabled and applies the "memory.high" control.
+type memoryController struct {
+	baseController
+	limit uint64
+}
+
+func (c memoryController) apply() error {
+	limit := strconv.FormatUint(c.limit, 10)
+	return errors.WithStack(c.baseController.apply(memoryHigh, limit))
+}
+
+// diskReadBpsController enables and appplies the rbps "io.max" control.
+type diskReadBpsController struct {
+	baseController
+	limit uint64
+}
+
+func (c diskReadBpsController) apply() error {
+	minors, err := readDiskDeviceMinors()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	for _, minor := range minors {
+		value := fmt.Sprintf("%d:%d rbps=%d", diskDevices, minor, c.limit)
+		if err := c.baseController.apply(ioMax, value); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+	return nil
+}
+
+// newDiskReadBpsController creates a diskReadBpsController instance.
+func newDiskReadBpsController(cgroup Cgroup, limit uint64) *diskReadBpsController {
+	return &diskReadBpsController{
+		baseController: baseController{name: io, cgroup: cgroup},
+		limit:          limit,
+	}
+}
+
+// newDiskWriteBpsController creates a diskWriteBpsController instance.
+func newDiskWriteBpsController(cgroup Cgroup, limit uint64) *diskWriteBpsController {
+	return &diskWriteBpsController{
+		baseController: baseController{name: io, cgroup: cgroup},
+		limit:          limit,
+	}
+}
+
+// diskReadBpsController enables and appplies the wbps "io.max" control.
+type diskWriteBpsController struct {
+	baseController
+	limit uint64
+}
+
+func (c diskWriteBpsController) apply() error {
+	minors, err := readDiskDeviceMinors()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	for _, minor := range minors {
+		value := fmt.Sprintf("%d:%d wbps=%d", diskDevices, minor, c.limit)
+		if err := c.baseController.apply(ioMax, value); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+	return nil
+}
+
+// baseController owns controller logic shared by most controller implementations.
+type baseController struct {
+	name   string
+	cgroup Cgroup
+}
+
+// enable enables a controller by writing to the cgroup.subtree_control file of
+// the cgroup.
+func (c baseController) enable() error {
+	file := path.Join(c.cgroup.path, cgroupSubtreeControl)
+	fd, err := os.OpenFile(file, os.O_WRONLY, fileMode)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer fd.Close()
```

Comment:
Ditto for deferred close, same elsewhere if applicable.

`os.WriteFile` could save you some work too.

Reply:
Updated